test/junitQueryTest/JUnitQueryTest.gmk

```diff
@@ -0,0 +1,246 @@
+
+#----------------------------------------------------------------------
+#
+# Execute tests when duplicates are given: last one wins (m13, all)
+# See related test for the reverse order
+
+$(BUILDTESTDIR)/JUnitQueryTest.m13.all.ok: $(JTREG_IMAGEDIR)/lib/javatest.jar \
+		$(JTREG_IMAGEDIR)/lib/jtreg.jar \
+		$(JTREG_IMAGEDIR)/bin/jtreg
+	$(RM) $(@:%.ok=%)
+	$(MKDIR) $(@:%.ok=%)
+	JT_JAVA=$(JDKHOME) JTHOME=$(JTREG_IMAGEDIR) \
+	    $(JTREG_IMAGEDIR)/bin/jtreg $(JTREG_OPTS) \
+		-w:$(@:%.ok=%)/work -r:$(@:%.ok=%)/report \
+		-jdk:$(JDKHOME) \
+		-va \
+		$(TESTDIR)/junitQueryTest/a/b/c/Test1.java?m13 \
+		$(TESTDIR)/junitQueryTest/a/b/c/Test1.java \
+		 > $(@:%.ok=%)/jt.log 2>&1
+	OUT=$$( $(ECHO) $$( $(AWK) '/STDOUT:/,/STDERR:/{ print $0; } { }' $(@:%.ok=%)/jt.log | $(GREP) -v STD ) ) ; \
+	if [ "$$OUT" != "Test1.m11 Test1.m12 Test1.m13" ]; then \
+	    echo "unexpected set of tests run: $$OUT"; exit 1 ; \
+	fi
+	$(GREP) "a/b/c/ containers: 3, .* started: 3, succeeded: 3" $(@:%.ok=%)/report/text/junit.txt
+	echo $@ passed at `date` > $@
+
+TESTS.jtreg += $(BUILDTESTDIR)/JUnitQueryTest.m13.all.ok
+
+#----------------------------------------------------------------------
+#
+# Execute tests when duplicates are given: last one wins (all, m13)
+# See related test for the reverse order
+
+$(BUILDTESTDIR)/JUnitQueryTest.all.m13.ok: $(JTREG_IMAGEDIR)/lib/javatest.jar \
+		$(JTREG_IMAGEDIR)/lib/jtreg.jar \
+		$(JTREG_IMAGEDIR)/bin/jtreg
+	$(RM) $(@:%.ok=%)
+	$(MKDIR) $(@:%.ok=%)
+	JT_JAVA=$(JDKHOME) JTHOME=$(JTREG_IMAGEDIR) \
+	    $(JTREG_IMAGEDIR)/bin/jtreg $(JTREG_OPTS) \
+		-w:$(@:%.ok=%)/work -r:$(@:%.ok=%)/report \
+		-jdk:$(JDKHOME) \
+		-va \
+		$(TESTDIR)/junitQueryTest/a/b/c/Test1.java \
+		$(TESTDIR)/junitQueryTest/a/b/c/Test1.java?m13 \
+		 > $(@:%.ok=%)/jt.log 2>&1
+	OUT=$$( $(ECHO) $$( $(AWK) '/STDOUT:/,/STDERR:/{ print $0; } { }' $(@:%.ok=%)/jt.log | $(GREP) -v STD ) ) ; \
+	if [ "$$OUT" != "Test1.m13" ]; then \
+	    echo "unexpected set of tests run: $$OUT"; exit 1 ; \
+	fi
+	$(GREP) "a/b/c/ containers: 3, .* started: 1, succeeded: 1" $(@:%.ok=%)/report/text/junit.txt
+	echo $@ passed at `date` > $@
+
+TESTS.jtreg += $(BUILDTESTDIR)/JUnitQueryTest.all.m13.ok
+
+#----------------------------------------------------------------------
+#
+# Invalid use of query
+
+$(BUILDTESTDIR)/JUnitQueryTest.invalidQuery.ok: $(JTREG_IMAGEDIR)/lib/javatest.jar \
+		$(JTREG_IMAGEDIR)/lib/jtreg.jar \
+		$(JTREG_IMAGEDIR)/bin/jtreg
+	$(RM) $(@:%.ok=%)
+	$(MKDIR) $(@:%.ok=%)
+	JT_JAVA=$(JDKHOME) JTHOME=$(JTREG_IMAGEDIR) \
+		$(JTREG_IMAGEDIR)/bin/jtreg $(JTREG_OPTS) \
+		-w:$(@:%.ok=%)/work -r:$(@:%.ok=%)/report \
+		-jdk:$(JDKHOME) \
+		-va \
+		$(TESTDIR)/junitQueryTest/a/b/c?m13 \
+			> $(@:%.ok=%)/jt.log 2>&1 || \
+		true "non-zero exit code from jtreg intentionally ignored"
+	$(GREP) -s "Error: Invalid use of query component: a/b/c?m13" $(@:%.ok=%)/jt.log
+	echo $@ passed at `date` > $@
+
+TESTS.jtreg += $(BUILDTESTDIR)/JUnitQueryTest.invalidQuery.ok
+
```

Comment:
I'm sure there are others, but behavior when the query is associated with a valid test springs to mind, i.e.

Negative test: a/b/c/Test1.java?m14 

Also test: a/b/c/Test1.java?m14 and a/b/c/Test.java

Reply:
I can add a negative test, such as for `?m14` but you just get an exception from JUnit about "method not found", so it counts as a "vanilla" test failure.

This would extend to the second proposed example: the tests in the file without a query would be executed, and the test with a bad query would fail.

Thinking about whether `jtreg` could detect and report an *error* instead of the test *failed*, I guess we _could_ check for the existence of the method in `JUnitAction` (i.e. within jtreg itself) but this would require that `jtreg` is both compiled and run with a JDK at least as new as the class being tested ... and that would be bad.

Once we are in the test runner or test code, the only options are to pass or fail -- i.e. return normally or throw an exception. We could carve out a special case similar to `jtreg.SkipException` but I'm not convinced it is worth the effort.   If the user gives a bad name, you get an exception.

```
TEST RESULT: Failed. Execution failed: `main' threw exception: org.junit.platform.commons.JUnitException: TestEngine with ID 'junit-jupiter' failed to discover tests
```

```
org.junit.platform.commons.JUnitException: TestEngine with ID 'junit-jupiter' failed to discover tests
Caused by: org.junit.platform.commons.JUnitException: MethodSelector [className = 'Test1', methodName = 'm14', methodParameterTypes = ''] resolution failed
	... 15 more
Caused by: org.junit.platform.commons.PreconditionViolationException: Could not find method with name [m14] in class [Test1].
	... 20 more
```